### PR TITLE
Raise exceptions in utils.set_environ

### LINKED ISSUE #4127: UnboundLocalError: local variable 'bypass' referenced before assignment
When using requests with GAE dev_appserver sandbox, this variable gets referenced before assignment when making requests to the FB graph API

UnboundLocalError: local variable 'bypass' referenced before assignment
https://github.com/requests/requests/blob/master/requests/utils.py#L680

### PR DESCRIPTION
**Pull request by mlcrazy**
fixes #4127

### CODE REVIEW

File: requests/utils.py [hunk 1]
```diff
@@ -615,15 +615,15 @@ def set_environ(env_name, value):
...
-    try:
+        try:
+            yield
+        finally:
+            if old_value is None:
+                del os.environ[env_name]
+            else:
+                os.environ[env_name] = old_value
+    else:
         yield
```
Comment (Lukasa) [inline requests/utils.py]: Hrm, I don't love the two separate `yield` statements here. Do you mind refactoring the old `finally` statement to invert the conditional on `if value is None: return` and to instead have `if value is not None` then do the dance with `old_value`.

Comment (mlcrazy) [reply requests/utils.py]: @Lukasa will set this refactor up and add a test

Comment (mlcrazy) [reply requests/utils.py]: @Lukasa refactored the finally statement, let me know if this is more what you would like to see there, also added tests

File: requests/utils.py [hunk 2]
```diff
@@ -612,18 +612,18 @@ def set_environ(env_name, value):
     the environment variable 'env_name'.
 
     If 'value' is None, do nothing"""
+
```
Comment (Lukasa) [inline requests/utils.py]: We don't need this newline.

File: requests/utils.py [hunk 3]
```diff
@@ -612,18 +612,18 @@ def set_environ(env_name, value):
...
     if value is not None:
-        old_value = os.environ.get(env_name)
+        old_value = os.environ.get(env_name, None)
```
Comment (Lukasa) [inline requests/utils.py]: `None` is implicit in `get`, you can leave it out.

File: requests/utils.py [hunk 4]
```diff
@@ -618,12 +618,11 @@ def set_environ(env_name, value):
     try:
         yield
     finally:
-        if value is None:
-            return
-        if old_value is None:
-            del os.environ[env_name]
-        else:
-            os.environ[env_name] = old_value
+        if value is not None:
```
Comment (nateprewitt) [inline requests/utils.py]: @Lukasa what are your thoughts on adding `value_changed = value is not None`, and using that single value for both conditionals? The fact these two checks are bound is implicit and has the potential to get out of sync as soon as one changes.

File: tests/test_utils.py [hunk 5]
```diff
@@ -651,4 +652,31 @@ def QueryValueEx(key, value_name):
...
+
+@pytest.mark.parametrize(
+    'env_name, value', (
+            ('no_proxy', '192.168.0.0/24,127.0.0.1,localhost.localdomain'),
+            ('no_proxy', None),
+            ('a_new_key', '192.168.0.0/24,127.0.0.1,localhost.localdomain'),
+            ('a_new_key', None),
+    ))
+def test_set_environ(env_name, value):
+    """Tests set_environ will set environ values and will restore the environ
```
Comment (nateprewitt) [inline tests/test_utils.py]: Could we end this with a period and either bring the `"""` up to the same line or wrap the text down? We don't have a hard cap at 80 characters in Requests.

File: tests/test_utils.py [hunk 6]
```diff
@@ -651,4 +652,31 @@ def QueryValueEx(key, value_name):
...
+    environ_copy = copy.deepcopy(os.environ)
+    with set_environ(env_name, value):
+        assert os.environ.get(env_name) == value
+
+    assert os.environ == environ_copy
+
+
+def test_set_environ_raises_exception():
+    """Tests set_environ will raise exceptions in context when the
+    value parameter is None
```
Comment (nateprewitt) [inline tests/test_utils.py]: `.`

File: requests/utils.py (hunk 4 above)
Comment (Lukasa) [reply requests/utils.py]: Works for me!

**Comment by mlcrazy** [comment]
@Lukasa @nateprewitt, did some commit fixups, not sure if I did the HISTORY.rst part correctly

**Comment by Lukasa** [comment]
There are conflicts in the AUTHORS.rst file. :smile:

**Pull request merged**